Feat(flex): Add modifiers for gap to flex component (#5270)
patternfly/patternfly: src/patternfly/layouts/Flex/examples/Flex.md

    Flex item
  {{/l-flex-item}}
  {{#> l-flex-item}}
    Flex item
  {{/l-flex-item}}
{{/l-flex}}
<br>
<h3>
  Gap inherits but can be overridden when nested - <code>.pf-l-gap-xl > .pf-l-gap-xs</code>
</h3>
{{#> l-flex l-flex--modifier="pf-m-gap-xl"}}
  {{#> l-flex l-flex--modifier="pf-m-gap-xs"}}
    {{#> l-flex-item}}
      Flex item
    {{/l-flex-item}}
    {{#> l-flex-item}}
      Flex item
    {{/l-flex-item}}
  {{/l-flex}}
  {{#> l-flex}}
    {{#> l-flex-item}}
      Flex item
    {{/l-flex-item}}
    {{#> l-flex-item}}
      Flex item
    {{/l-flex-item}}
    {{#> l-flex-item}}
      Flex item
    {{/l-flex-item}}
  {{/l-flex}}
{{/l-flex}}
<br>
<h3>
  REMOVE AFTER TESTING Nested flex and items
</h3>
{{#> l-flex l-flex--modifier="pf-m-gap-xs"}}
  {{#> l-flex l-flex--modifier="pf-m-gap-xl"}}
    {{#> l-flex-item}}
      Flex item
    {{/l-flex-item}}
    {{#> l-flex-item}}
      Flex item
    {{/l-flex-item}}
  {{/l-flex}}

  {{#> l-flex-item}}
    Flex item
  {{/l-flex-item}}
  {{#> l-flex-item}}
    Flex item
  {{/l-flex-item}}

  {{#> l-flex l-flex--modifier="pf-m-gap-xs"}}
    {{#> l-flex-item}}
      Flex item
    {{/l-flex-item}}
  {{/l-flex}}
{{/l-flex}}
<br>

<h3>

>   Gap with individually spaced items - <code>.pf-m-gap-xs > pf-m-spacer-{xs,sm,md,lg,xl,2xl,3xl}</code>.

Reviewer: oops! Forgot to submit this comment in my previous review

This works if the spacer is larger than the gap, but if the gap is larger than the spacer, the item will have a negative margin, causing it to go out of bounds.

<img width="541" alt="Screenshot 2022-12-08 at 6 32 48 PM" src="https://user-images.githubusercontent.com/35148959/206596711-052b6c48-c938-47aa-a476-da386f809d8a.png">

</h3>
{{#> l-flex l-flex--modifier="pf-m-gap-xs"}}
  {{#> l-flex-item}}
    Item
  {{/l-flex-item}}
  {{#> l-flex-item}}
    Item
  {{/l-flex-item}}
  {{#> l-flex-item l-flex-item--modifier="pf-m-spacer-none"}}
    Item - none
  {{/l-flex-item}}
  {{#> l-flex-item l-flex-item--modifier="pf-m-spacer-xs"}}
    Item - xs
  {{/l-flex-item}}
  {{#> l-flex-item l-flex-item--modifier="pf-m-spacer-sm"}}
    Item - sm
  {{/l-flex-item}}
  {{#> l-flex-item l-flex-item--modifier="pf-m-spacer-md"}}
    Item - md
  {{/l-flex-item}}
  {{#> l-flex-item l-flex-item--modifier="pf-m-spacer-lg"}}
    Item - lg
  {{/l-flex-item}}
  {{#> l-flex-item l-flex-item--modifier="pf-m-spacer-xl"}}
    Item - xl
  {{/l-flex-item}}
  {{#> l-flex-item l-flex-item--modifier="pf-m-spacer-2xl"}}
    Item - 2xl
  {{/l-flex-item}}
  {{#> l-flex-item l-flex-item--modifier="pf-m-spacer-3xl"}}
    Item - 3xl
  {{/l-flex-item}}
{{/l-flex}}
<br>
<h3>

>   Space items will override column gap - <code>.pf-m-gap-xs.pf-m-space-items-xl</code>

Reviewer: What's the use case for using them together? On the main axis, it continues to use gap, but offsets the gap with the margin specified in the `pf-m-space-items` class, so you end up with negative margins when the gap is larger than the margin here, too.

<img width="685" alt="Screenshot 2022-12-08 at 6 35 36 PM" src="https://user-images.githubusercontent.com/35148959/206595967-c48a8822-dee9-44d9-9b72-be444c6d0b01.png">

  {{#> l-flex-item}}
    Flex item
  {{/l-flex-item}}
  {{#> l-flex-item}}
    Flex item
  {{/l-flex-item}}
  {{#> l-flex-item}}
    Flex item
  {{/l-flex-item}}
{{/l-flex}}
<br>
<h3>
  Column gap using <code>.pf-m-column-gap</code>
</h3>
{{#> l-flex l-flex--modifier="pf-m-column-gap-lg"}}
  {{#> l-flex-item}}
    Flex item
  {{/l-flex-item}}
  {{#> l-flex-item}}
    Flex item
  {{/l-flex-item}}
  {{#> l-flex-item}}
    Flex item
  {{/l-flex-item}}
  {{#> l-flex-item}}
    Flex item
  {{/l-flex-item}}
  {{#> l-flex-item}}
    Flex item
  {{/l-flex-item}}
  {{#> l-flex-item}}
    Flex item
  {{/l-flex-item}}
{{/l-flex}}
<br>
<h3>
  Gap inherits, but can be overridden when nested - <code>.pf-l-gap-xl > .pf-l-gap-xs</code>
</h3>
{{#> l-flex l-flex--modifier="pf-m-gap-xl"}}
  {{#> l-flex l-flex--modifier="pf-m-gap-xs"}}
    {{#> l-flex-item}}
      Flex item
    {{/l-flex-item}}
    {{#> l-flex-item}}
      Flex item
    {{/l-flex-item}}
  {{/l-flex}}
  {{#> l-flex}}
    {{#> l-flex-item}}
      Flex item
    {{/l-flex-item}}
    {{#> l-flex-item}}
      Flex item
    {{/l-flex-item}}
    {{#> l-flex-item}}
      Flex item
    {{/l-flex-item}}
  {{/l-flex}}
{{/l-flex}}
<br>

<h3>

>   Row gap with individually spaced items - <code>.pf-m-row-gap-xs > pf-m-spacer-{xs,sm,md,lg,xl,2xl,3xl}</code>. Using `.pf-m-spacer` is not recommended with column gap.

Reviewer: nit. 

```suggestion
  Row gap with individually spaced items - <code>.pf-m-row-gap-xs > pf-m-spacer-{xs,sm,md,lg,xl,2xl,3xl}</code>. Using <code>.pf-m-spacer</code> is not recommended with column gap.
```

also fwiw the class `.ws-code` formats these code blocks a bit better, though I don't know what the solution is there. I noticed in our a11y docs we're hard coding those classes when we write out the element names, which doesn't seem ideal. Maybe we need to target elements with better selectors to style these elements when we have to write out the tag, or maybe a way to be able to include markdown in these blocks so it gets added automagically? https://github.com/patternfly/patternfly-org/blob/main/packages/v4/patternfly-docs/content/accessibility/card/card.md

Added this org issue to track - https://github.com/patternfly/patternfly-org/issues/3311

>   Row gap with individually spaced items - <code>.pf-m-row-gap-xs > pf-m-spacer-{xs,sm,md,lg,xl,2xl,3xl}</code>. Using `.pf-m-spacer` is not recommended with column gap.

Reviewer: nit. 

```suggestion
  Row gap with individually spaced items - <code>.pf-m-row-gap-xs > pf-m-spacer-{xs,sm,md,lg,xl,2xl,3xl}</code>. Using <code>.pf-m-spacer</code> is not recommended with column gap.
```

also fwiw the class `.ws-code` formats these code blocks a bit better, though I don't know what the solution is there. I noticed in our a11y docs we're hard coding those classes when we write out the element names, which doesn't seem ideal. Maybe we need to target elements with better selectors to style these elements when we have to write out the tag, or maybe a way to be able to include markdown in these blocks so it gets added automagically? https://github.com/patternfly/patternfly-org/blob/main/packages/v4/patternfly-docs/content/accessibility/card/card.md

Added this org issue to track - https://github.com/patternfly/patternfly-org/issues/3311

  {{#> l-flex-item}}
    Item
  {{/l-flex-item}}
  {{#> l-flex-item}}
    Item
  {{/l-flex-item}}
  {{#> l-flex-item l-flex-item--modifier="pf-m-spacer-none"}}
    Item - none
  {{/l-flex-item}}
  {{#> l-flex-item l-flex-item--modifier="pf-m-spacer-xs"}}
    Item - xs
  {{/l-flex-item}}
  {{#> l-flex-item l-flex-item--modifier="pf-m-spacer-sm"}}
    Item - sm
  {{/l-flex-item}}
  {{#> l-flex-item l-flex-item--modifier="pf-m-spacer-md"}}
    Item - md
  {{/l-flex-item}}
  {{#> l-flex-item l-flex-item--modifier="pf-m-spacer-lg"}}
    Item - lg
  {{/l-flex-item}}
  {{#> l-flex-item l-flex-item--modifier="pf-m-spacer-xl"}}
    Item - xl
  {{/l-flex-item}}
  {{#> l-flex-item l-flex-item--modifier="pf-m-spacer-2xl"}}
    Item - 2xl
  {{/l-flex-item}}
  {{#> l-flex-item l-flex-item--modifier="pf-m-spacer-3xl"}}
    Item - 3xl
  {{/l-flex-item}}
{{/l-flex}}
```
### Usage
| Class | Applied to | Outcome |
| -- | -- | -- |
| `.pf-m-gap-{none, xs, sm, md, lg, xl, 2xl}{-on-[breakpoint]}`, `.pf-m-row-gap-{none, xs, sm, md, lg, xl, 2xl}{-on-[breakpoint]}`, `.pf-m-column-gap-{none, xs, sm, md, lg, xl, 2xl}{-on-[breakpoint]}` | `.pf-l-flex` | Creates a gap, row gap, or column gap. |

> Note: Using the spacer system (`.pf-m-spacer` or `.pf-m-item-space-items`) with gap or column gap is not recommended, as it could cause elements to extend past the parent's boundary in the inline direction.

Reviewer: nit - I think that's probably more clear, since the main axis refers to the flex direction.

```suggestion
Note: Using the spacer system (`.pf-m-spacer` or `.pf-m-item-space-items`) with gap or column gap is not recommended, as it could cause elements to extend past the parent's boundary along the main axis.
```

<br/>
<h3>
  Row gap using <code>.pf-m-row-gap-sm</code>
</h3>
{{#> l-flex l-flex--modifier="pf-m-row-gap-sm"}}
  {{#> l-flex-item}}
    Flex item
  {{/l-flex-item}}
  {{#> l-flex-item}}
    Flex item
  {{/l-flex-item}}
  {{#> l-flex-item}}
    Flex item
  {{/l-flex-item}}
  {{#> l-flex-item}}
    Flex item
  {{/l-flex-item}}
  {{#> l-flex-item}}
    Flex item
  {{/l-flex-item}}
  {{#> l-flex-item}}
    Flex item
  {{/l-flex-item}}
  {{#> l-flex-item}}
    Flex item
  {{/l-flex-item}}
  {{#> l-flex-item}}
    Flex item
  {{/l-flex-item}}
  {{#> l-flex-item}}
    Flex item
  {{/l-flex-item}}
  {{#> l-flex-item}}
    Flex item
  {{/l-flex-item}}
  {{#> l-flex-item}}
    Flex item
  {{/l-flex-item}}
  {{#> l-flex-item}}
    Flex item
  {{/l-flex-item}}
{{/l-flex}}
<br>
<h3>
  Column gap using <code>.pf-m-column-gap</code>
</h3>
{{#> l-flex l-flex--modifier="pf-m-column-gap-sm"}}
  {{#> l-flex-item}}
    Flex item
  {{/l-flex-item}}
  {{#> l-flex-item}}
    Flex item
  {{/l-flex-item}}
  {{#> l-flex-item}}
    Flex item
  {{/l-flex-item}}
  {{#> l-flex-item}}
    Flex item
  {{/l-flex-item}}
  {{#> l-flex-item}}
    Flex item
  {{/l-flex-item}}
  {{#> l-flex-item}}
    Flex item

>   {{/l-flex-item}}

Reviewer: Would you mind adding a few more so the row wraps by default and the column gap is visible? The first two examples use 12 items, that would probably make sense to keep them consistent.

Author: Done. I think the column modifier works correctly now with gap. I think long term there are some improvements we could do to use gap by default for both directions.

    Item - none
  {{/l-flex-item}}
  {{#> l-flex-item l-flex-item--modifier="pf-m-spacer-xs"}}
    Item - xs
  {{/l-flex-item}}
  {{#> l-flex-item l-flex-item--modifier="pf-m-spacer-sm"}}
    Item - sm
  {{/l-flex-item}}
  {{#> l-flex-item l-flex-item--modifier="pf-m-spacer-md"}}
    Item - md
  {{/l-flex-item}}
  {{#> l-flex-item l-flex-item--modifier="pf-m-spacer-lg"}}
    Item - lg
  {{/l-flex-item}}
  {{#> l-flex-item l-flex-item--modifier="pf-m-spacer-xl"}}
    Item - xl
  {{/l-flex-item}}
  {{#> l-flex-item l-flex-item--modifier="pf-m-spacer-2xl"}}
    Item - 2xl
  {{/l-flex-item}}
  {{#> l-flex-item l-flex-item--modifier="pf-m-spacer-3xl"}}
    Item - 3xl
  {{/l-flex-item}}
{{/l-flex}}
```
### Usage
Use gap modifiers to space flex items evenly. If individual spacing of items along the main axis is needed, the [flex spacer system](#spacer) can be used instead of row gap.

Using the spacer system (`.pf-m-spacer` or `.pf-m-item-space-items`) with gap or column gap is not recommended, as in some cases, it could cause elements to extend past the parent's boundary along the main axis.

| Class | Applied to | Outcome |
| -- | -- | -- |

> | `.pf-m-gap-{none, xs, sm, md, lg, xl, 2xl}{-on-[breakpoint]}`, `.pf-m-row-gap-{none, xs, sm, md, lg, xl, 2xl}{-on-[breakpoint]}`, `.pf-m-column-gap-{none, xs, sm, md, lg, xl, 2xl}{-on-[breakpoint]}` | `.pf-l-flex` | Creates a gap, row gap, or column gap. |

Reviewer: I think these should be unique rows in the table for `gap`, `row-gap`, and `column-gap`.

Also since you don't have to specify a size, I think the class suffix should be `.pf-m-gap{-{none, xs, sm, md, lg, xl, 2xl}{-on-[breakpoint]}}`, or something like that, just denoting that `.pf-m-[row/column-]gap` can be used on its own. @mattnolting wdyt, I feel like you know the curly vs regular braces syntax better than I do

> | `.pf-m-gap-{none, xs, sm, md, lg, xl, 2xl}{-on-[breakpoint]}`, `.pf-m-row-gap-{none, xs, sm, md, lg, xl, 2xl}{-on-[breakpoint]}`, `.pf-m-column-gap-{none, xs, sm, md, lg, xl, 2xl}{-on-[breakpoint]}` | `.pf-l-flex` | Creates a gap, row gap, or column gap. |

Reviewer: Needs the breakpoint link.

```suggestion
| `.pf-m-gap-{none, xs, sm, md, lg, xl, 2xl}{-on-[breakpoint]}`, `.pf-m-row-gap-{none, xs, sm, md, lg, xl, 2xl}{-on-[breakpoint]}`, `.pf-m-column-gap-{none, xs, sm, md, lg, xl, 2xl}{-on-[breakpoint]}` | `.pf-l-flex` | Creates a gap, row gap, or column gap at optional [breakpoint](https://www.patternfly.org/v4/developer-resources/global-css-variables#breakpoint-variables-and-class-suffixes). |
```

Author: This isn't on any of the individual modifiers throughout this page because there are so many, but that link is mentioned at the top of the page in a breakpoint section.

  {{/l-flex-item}}
  {{#> l-flex-item}}
    Flex item
  {{/l-flex-item}}
  {{#> l-flex-item}}
    Flex item
  {{/l-flex-item}}
  {{#> l-flex-item}}
    Flex item
  {{/l-flex-item}}
  {{#> l-flex-item}}
    Flex item
  {{/l-flex-item}}
  {{#> l-flex-item}}
    Flex item
  {{/l-flex-item}}
  {{#> l-flex-item}}
    Flex item
  {{/l-flex-item}}
  {{#> l-flex-item}}
    Flex item
  {{/l-flex-item}}
{{/l-flex}}
<br>
<h3>
  Column gap using <code>.pf-m-column-gap</code>
</h3>
{{#> l-flex l-flex--modifier="pf-m-column-gap-sm"}}
  {{#> l-flex-item}}
    Flex item
  {{/l-flex-item}}
  {{#> l-flex-item}}
    Flex item
  {{/l-flex-item}}
  {{#> l-flex-item}}
    Flex item
  {{/l-flex-item}}
  {{#> l-flex-item}}
    Flex item
  {{/l-flex-item}}
  {{#> l-flex-item}}
    Flex item
  {{/l-flex-item}}
  {{#> l-flex-item}}
    Flex item
  {{/l-flex-item}}
  {{#> l-flex-item}}
    Flex item
  {{/l-flex-item}}
  {{#> l-flex-item}}
    Flex item
  {{/l-flex-item}}
  {{#> l-flex-item}}
    Flex item
  {{/l-flex-item}}
  {{#> l-flex-item}}
    Flex item
  {{/l-flex-item}}
  {{#> l-flex-item}}
    Flex item
  {{/l-flex-item}}
  {{#> l-flex-item}}
    Flex item
  {{/l-flex-item}}
{{/l-flex}}
<br>
<h3>

>   Gap inherits, but can be overridden when nested - <code>.pf-l-gap-xl > .pf-l-gap-xs</code>

Reviewer: I'm probably not reading it right, but not sure what "gap inherits" is referring to. The `xl` gap only applies to the parent layout. If you remove the nested gap (no space modifer, uses the default spacer system), or just use `.pf-m-gap`, the default spacers are used.

>   Gap inherits, but can be overridden when nested - <code>.pf-l-gap-xl > .pf-l-gap-xs</code>

Reviewer: ```suggestion
  Gap inherits, but can be overridden when nested - <code>.pf-m-gap-xl > .pf-m-gap-xs</code>
```
